Rust bindings refactor

emesare opened the pull request:
Opening this PR for public discussion. This needs more testing and also lacks a few major changes that make sense to land alongside the already existing set of breaking changes.

The Rust bindings are being heavily refactored for the 4.3 release. This is to address many of the shortcomings and historical issues, namely around maintaining and safety.

Some of the additions of this branch include:
- Revised project structure
- Consistent formatting (with `rustfmt`)
- Many UB fixes
- Unit tests
- More/updated documentation 
- Real rust examples (i.e. `cargo run --example`)
- Many of the already opened PR's
- Additional APIs
- Updated and idiomatic linking to the core
- Consolidation of API's (removal of historical API's)
- Rust main thread handler
- Type driven API's (ex. `RegisterId`)
- Updated CI

# Reviewer Guide
**TODO**
- Explain the CI and add a bunch of links to relevant GitHub Actions documentation
- Explain the project structure (this isn't really necessary for consumers)
- Explain the usage of Ref
- Explain the usage of BnString (and strings in general)
- Explain how to review PR's
- Explain how to run tests

rbran committed:
improve the CustomBinaryView init process

rbran committed:
Add more HLIL related functions to rust

rbran committed:
Implement rust TypeParser

rbran committed:
fix Platform TypeParser related functions

rbran committed:
separate User and System implementations of TypeParserResult

rbran committed:
Implement rust TypeContainer

rbran committed:
impl rust Collaboration and Remote API

emesare committed:
Fix typo

emesare committed:
Link core in rust

When linking you must depend on the -sys crate.

This is because linker arguments (what says where to find binaryninjacore) are NOT transitive. The top level application crate MUST provide it.

For more information see:
- https://github.com/rust-lang/cargo/issues/9554#issuecomment-857882964
- https://github.com/oxidecomputer/omicron/pull/225

emesare committed:
Canonicalize path in `create_delete_empty` test

emesare committed:
Cleanup rust api

Moves a bunch of stuff out of src/types.rs that did not belong:
- Confidence
- Variable
- Function specific stuff

- Refactored InstructionInfo, see the msp430 and riscv examples.
- Renamed Function::from_raw to Function::ref_from_raw and fixed places where the ref was incremented twice
- Fixed FunctionRecognizer leaking functions (see above)
- Fixed some apis incorrectly returning Result where Option is clearer
- Started to move destructured types to the From trait for converting to an from ffi types, see Location for an example
- Started to remove bad module level imports (importing std modules like mem all over the place)
- Moved some wrapper handle types to named handle field (this improves readability), see CoreArchitecture for an example
- Removed some unintuitive getters, this is bad practice for Rust code, just access the field directly, see DataVariable for an example
- General code cleanup, purposely did not run rustfmt, that will be a single seperate commit

emesare committed:
More rust cleanup

- Fixed invalid views being able to invoke UB when dealing with databases
- Cleaned up some helper code in dwarf_import
- Fixed inverted is_null checks causing crashes! Oops!

emesare committed:
More rust cleanup

Still a WIP, I think branch info is still invalid, need to figure out the issue there.

- Fixed some invalid Ref lifetimes when constructing indirectly, see Array<DataVariable> for example
- Added some more comments
- Removed some "magic" functions like MLIL Function::ssa_variables

There are still a bunch of invalid lifetimes that aren't crashing us due to the usage of those API's not living long enough. But they ARE an issue.

emesare committed:
More rust cleanup

Trying to comment more TODO's as I go along.

- Renamed llil::Function to llil::LowLevelILFunction for clarity and consistency
- Take base structures by ref in StructureBuilder::set_base_structures to prevent premature drops
- Added more raw to wrapper conversions
- Removed UB prone apis
- Getting rid of more core module references, use std!
- Removed improper Guard usage in array impls for wrapper types with no context
- Untangling the UB of the Label api, still only half done :/

emesare committed:
More rust cleanup

- Misc formatting
- Made Logger ref counted
- Fixed leaking name of logger every time something was logged
- Fixed the last (hopefully) of the unresolved labels
- Simplified some code

emesare committed:
Fix leak in DebugInfo::AddType

componentArray was never freed

emesare committed:
More rust cleanup

emesare committed:
Remove vendored pdb crate

Use cargo to manage the git repo ref instead

emesare committed:
Fix misc rustdoc warnings

emesare committed:
Move actual plugins out of `rust/examples` and into `plugins`

This is where all shipped public plugins that are not arch/view/platform/lang will be at from now on

Originally they were in the rust workspace, meaning they all shared a Cargo.lock which is ill-advised.

emesare committed:
More rust cleanup

- More clarification on plugin/executable requirements
- Made examples actually rust examples
- Add Display impl for InstructionTextToken
- Renamed feature "noexports" to "no_exports"

emesare committed:
Move under_construction.png to assets directory

This really did bother me

emesare committed:
Remove unneeded `extern crate bindgen`

emesare committed:
Replace nop'd log::info with println

We don't register a compatible log sink so they will just get sent into the void

emesare committed:
Move inline tests into tests directory

This is being done in the hopes of curbing the multi-thousand line files that will occur once we flesh out the tests

emesare committed:
Format rust code

emesare committed:
Update rust ci

Still need to add support for running tests in ci

emesare committed:
More rust cleanup

- Architecture id's are now typed accordingly
- Fix some clippy lints
- Make instruction index public in LLIL
- Removed useless helper functions
- LLIL expressions and instruction indexes are now typed accordingly

emesare committed:
Generate binaryninjacore-sys documentation

This should show binaryninjacore-sys alongside binaryninja crate

emesare committed:
More rust cleanup

- Remove lazy_static dependency
- Remove hacky impl Debug for Type and just use the display impl
- Add more debug impls
- Reorder some top level namespace items
- Add first type test

emesare committed:
Remove unneeded script helper in rust api

emesare committed:
More rust cleanup

- Added main thread handler api
- Register a headless main thread handler by default in initialization
- Refactor QualifiedName to be properly owned
- Loosened some type constraints on some apis involving QualifiedName
- Fixed some apis that were crashing due to incorrect param types
- Removed extern crate cruft for log crate
- Simplified headless initialization using more wrapper apis
- Fixed segments leaking because of no ref wrapper, see BinaryViewExt::segment_at
- Added rstest to manage headless init in unit tests
- Added some more unit tests
- Refactored demangler api to be more ergonomic
- Fixed minidump plugin not building

emesare committed:
More rust cleanup

- Fixup usage of QualifiedName in plugins
- Make QualifiedName more usable

emesare committed:
More rust cleanup

- Hopefully fixed the rust.yml CI
- Added TypePrinter API (this is still WIP and will crash)
- Added TypeParser API
- Added TypeContainer API
- More work around QualifiedName apis

emesare committed:
Oh your suppose to do this

emesare committed:
Add workflow_dispatch trigger to rust.yml

emesare committed:
More rust fixes

- Swapped some usage of raw 255 to MAX_CONFIDENCE, no one likes magic numbers
- New InstructionTextToken API, complete with owned data, this still needs a lot of testing.
- InstructionTextTokenKind describes a destructured InstructionTextToken, this should make token usage much clearer, some docs pending
- Added some misc Default and Debug impls
- Updated TypePrinter and architectures to use new InstructionTextToken API

emesare committed:
Misc formatting changes

emesare committed:
More rust cleanup

- Fixed MANY memory leaks (most due to QualifiedName)
- Made StructureBuilder more builder and less structure
- Fixed CMakeLists.txt that were globbing entire api, resulting in 100 second slowdown on cmake generation
- Added more Debug impl's
- Added some more tests
- Fixed TypeParserResult UB
- Moved the From impls to blank impl for clarity, we have multiple different variants of core to rust for some structures, it should be explicit which one you are choosing.
- PossibleValueSet should now be able to allocate so we can go from rust to core with those variants that require allocation
- Misc doc code formatting

emesare committed:
Misc clippy lints and clippy CI

Co-authored-by: Michael Krasnitski <michael.krasnitski@gmail.com>

emesare committed:
Fix typo in rust CI

emesare committed:
Misc rust formatting

emesare commented:
Documentation needs to be updated to reflect that linking to the core will no longer be optional for plugins. This should be done proactively before we start versioning, because once that is done there _will_ be a strict requirement to depend directly on the `-sys` crate.

emesare committed:
Fix misc typos and add typos to rust CI

emesare commented:
Mentioning people who have code that will be moved with this PR, if you have unmerged code (i.e. a branch with fixes) please let me know and I can try to incorporate the changes alongside this PR. Otherwise you shouldn't need to do anything.

@CouleeApps The pdb import plugin was moved to `/plugins`
@negasora The dwarf plugins were moved to `/plugins`
@rbran The idb import plugin was moved to `/plugins`

emesare commented:
This PR will also close the following:
- https://github.com/Vector35/binaryninja-api/pull/6243
- https://github.com/Vector35/binaryninja-api/pull/5898
- https://github.com/Vector35/binaryninja-api/pull/5671
- https://github.com/Vector35/binaryninja-api/pull/5584
- https://github.com/Vector35/binaryninja-api/pull/5431
- https://github.com/Vector35/binaryninja-api/pull/5465
- https://github.com/Vector35/binaryninja-api/pull/6257
- https://github.com/Vector35/binaryninja-api/pull/5773

emesare committed:
Add cargo workspace

This will help tooling and external contributors get a map of the rust crates within binaryninja-api

emesare committed:
More rust cleanup

- Format all rust plugins
- Fix some tests that were out of date
- Simplify WARP tests to only binaries, building object files from source is a pain
- Link to core in all rust plugins
- Fix some memory leaks
- Update warp insta snapshots
- Fix some misc clippy lints

emesare committed:
Run rust tests in CI

This commit also coincides with the creation of the "testing" environment which exposes a BN_SERIAL secret for pulling a headless Binary Ninja

emesare committed:
Install missing wayland dependency in github CI

Apparently its needed for linux file picker for the WARP integration

emesare committed:
Set the BINARYNINJADIR so rust can find binaryninjacore in CI

The chances of this working are low

emesare committed:
Misc remove unused dependency

emesare committed:
Rust misc formatting fixes

emesare committed:
Improve initialization in rust headless scripts

Provide sensible errors and validation to rust headless scripts, solves https://github.com/Vector35/binaryninja-api/issues/5796

emesare committed:
Add BN_LICENSE environment variable to rust CI

We pass the serial to download binary ninja, but we never provided the license for core initialization

emesare committed:
Fix typo

mkrasnitski commented:
How come the jobs are duplicated? Seems like `push` and `pull_request` workflows would be mutually exclusive.

emesare commented:
> How come the jobs are duplicated? Seems like `push` and `pull_request` workflows would be mutually exclusive.

They aren't mutually exclusive, we might want to restrict `push` to just the main branch, not entirely sure what the best way to go about it is.

emesare committed:
More rust cleanup

- Improved binary view initialization (see init_with_opts)
- Allow floating license to free itself before initialization
- Add initialization unit test
- Add better Debug impls for some common types
- Use Path api for opening binary views, this is not breaking as it uses the AsRef impl
- Bump rayon dependency and constrain dependencies to x.x

emesare committed:
Update readme and include that directly in the rustdocs

emesare committed:
More rust documentation changes

emesare committed:
Add format comment to InitializationOptions::with_license

emesare committed:
Misc formatting and clippy lint allow

emesare committed:
More rust cleanup

- Remove under_construction.png from the build.rs it has been removed
- Use InstructionIndex in more places
- Add missing PartialOrd and Ord impls for id types

emesare commented:
We need to add a doc file explaining the core linkage more in depth, and how to control the ABI version used independently of the rust API. 

I added an example here https://github.com/emesare/simple_rust_binja/blob/master/Cargo.toml but obviously needs to be expanded on and explained.

We should also have a simple "best practices" that is a TLDR, but basically boils down to, please pin your code to a commit or branch, or something other than "dev".

emesare commented:
As for can we document what API's are going to be changing, I don't have an exhaustive list, but in the case of a user wanting to know if they should use the rust api for a certain task, over lets say, C++ or python, It would be nice to have.

For an example: User wants to use the flow graph API, however there is not a consumer of that necessarily (other than dwarfdump) so the API might change once there is a consumer. 

The easiest way to determine if an API is subject to many changes is to see if there is a consumer. If there is no consumer we could document, either at the module level, or in some documentation that said API would be "unstable". I don't _really_ like doing it the other way around, that is, saying that an API is stable.

emesare commented:
The structure grouper is failing for the pdb import plugin, I am not entirely sure what I did to cause that, but the effect in a real binary is palpable.

Test: https://github.com/Vector35/binaryninja-api/actions/runs/12780281192/job/35626277295

The issue in a real binary:
![image](https://github.com/user-attachments/assets/396703cf-5c46-49c0-abb4-ce376726300d)

Left is before, and right is this branch

emesare commented:
We have a real issue with code coverage in the rust api, a large portion of the api is not used at all.

The solution is unit tests and more examples, which is what this PR is working towards. However measuring that is difficult, we might want to look into including some code coverage tool in CI.

- https://lib.rs/crates/cargo-llvm-cov
- https://github.com/xd009642/tarpaulin
- https://github.com/mozilla/grcov

emesare commented:
What is stopping this PR from being merged?

1. Some tests are failing for pdb import
2. Need to run through some binaries again to make sure there are no regressions performance-wise, this was done after some changes to the architecture api but has not been done again

emesare commented:
> As for can we document what API's are going to be changing, I don't have an exhaustive list, but in the case of a user wanting to know if they should use the rust api for a certain task, over lets say, C++ or python, It would be nice to have.
> 
> For an example: User wants to use the flow graph API, however there is not a consumer of that necessarily (other than dwarfdump) so the API might change once there is a consumer.
> 
> The easiest way to determine if an API is subject to many changes is to see if there is a consumer. If there is no consumer we could document, either at the module level, or in some documentation that said API would be "unstable". I don't _really_ like doing it the other way around, that is, saying that an API is stable.

For this I think both a little foreword in the README.md explaining what API's are actively being used, and therefor more reliable and better suited for consumers that are not comfortable with API's that weren't tested/or API's with the possibility of breaking changes. Aside from the README.md we might also want to add a warning on modules like MLIL and HLIL that they are subject to breaking changes.

The goal for this is to make it clear for those that do not want the API to change as frequently, if rust fits the bill, or if they should possibly use C++/Python api instead.

emesare commented:
Are there any more planned API changes for this PR?

No, I don't have anything else planned, and the changes that are occurring are on functions that previously had _no_ visible consumer.

emesare committed:
More rust cleanup

- Make workflow cloning explicit
- Add workflow tests
- Add missing property string list getting for settings
- Remove IntoActivityName (see https://github.com/Vector35/binaryninja-api/pull/6257)

emesare commented:
GitHub has no easy way to set deployment retention, this is insane behavior.

https://github.com/Vector35/binaryninja-api/deployments/testing

Every push to the rust code will trigger this, we might need to setup _another_ workflow to run on some arbitrary schedule to delete old deployments.

emesare committed:
More rust cleanup

This commit is half done

emesare commented:
<img width="1175" alt="image" src="https://github.com/user-attachments/assets/845e3d81-8b2f-4ab0-b5c8-914ba54fee00" />

Left is this PR, right is current dev. Any match that is traversing the IL fails. (Because the IL traversal is borked)

At some point during the last few days, IL traversal has been borked. Need to fix that obviously before this is merged.

emesare committed:
Misc rust formatting

emesare commented:
Next priority (due to the above) is to add LLIL unit tests. We need to have the `visit_tree` functions tested thoroughly to prevent regressions.

Most likely the WARP is not visiting the expressions, so first look at the `visit_tree` set of functions.

emesare committed:
More rust cleanup

- Renamed common name conflictions (I will put my justification in the PR)
- Fixed invalid instruction retrieval for LLIL
- Added common aliases for llil function, instruction and expression types (see my comment in the PR)
- Refactored the instruction retrieval for LLIL, MLIL and HLIL
- Added instruction index types to MLIL and HLIL
- Moved llil module to lowlevelil module (mlil and hlil will be moved as well)
- Added preliminary LLIL unit testing

emesare commented:
> Next priority (due to the above) is to add LLIL unit tests. We need to have the `visit_tree` functions tested thoroughly to prevent regressions.
> 
> Most likely the WARP is not visiting the expressions, so first look at the `visit_tree` set of functions.

This has been fixed with https://github.com/Vector35/binaryninja-api/pull/6309/commits/2b7557dee756b76f034cf00645585936f958e03e

emesare committed:
Fix typos

emesare committed:
Misc clippy fixes

emesare commented:
Noting that we need to move the mlil and hlil modules to their respective full name counterparts

emesare commented:
We still need more coverage of the llil api, mostly regarding the visitors.

The MLIL and HLIL api's I want to mark as unstable in the docs because I will need a second pass for those, I don't have time to rework them. Specifically so that we can rewrite MLIL and HLIL functions.

The LLIL lifter needs to be split apart and also have the Lifted restriction removed, we can emit regular LLIL through function workflows, meaning we also probably need to rename that part of the LLIL API to something other than "lifting", I am partial to something like "emitter" or something.

emesare commented:
> Are there any more planned API changes for this PR?
> 
> No, I don't have anything else planned, and the changes that are occurring are on functions that previously had _no_ visible consumer.

Clearly this is no longer the case :/

I wish I could stop changing the API but I know it will be a while before we can break API's again, so...

emesare commented:
I need to add a plugin example, I just don't know where to put it. I really dislike using the "examples" directory for something that isnt actually a rust example. I think a plugin_example subcrate (same directory placement of binaryninjacore-sys) would be the most appropriate. We need **_ONE_** plugin example, not 10. Moving forward we write application examples, people want examples they can just run, not something that requires them to build it and open binary ninja and figure out what the example is trying to do. They need to be self contained.

emesare commented:
We should also do a once over of the old rust_break_everything branch to make sure there wasn't any missing commits that were not rolled up into this branch.

emesare commented:
On the topic of imports:

- I am going to normalize all module names (i.e. `debuginfo` will become `debug_info`)
- I am _not_ going to add a prelude module

For why I am deciding against adding a prelude see [this reddit post](https://www.reddit.com/r/rust/comments/q6stlt/massive_preludes_why/?chainedPosts=t3_a7pcp2). Namely, we have very generic names that would end up existing within the prelude, things such as `Type` and `Database`, either we assume that our library is the only library a large portion of users will import with those names, in which case a prelude would be fine, _or_ we don't make that assumption.

If anyone disagrees with this please say something!

emesare committed:
More rust cleanup

- Normalized modules
- Split some code out into own files
- Fixed some UB in type archive and projects
- Improved API around type archives and projects substantially
- Added ProgressExecutor abstraction for functions which have a progress callback
- Improved background task documentation and added unit tests
- Added worker thread api and unit tests
- Moved some owned types to ref types, this is still not complete, but this is the path forward.
- Add external location/library accessors to the binary view
- Added some misc documentation
- Replaced mod.rs with the module name source file

Still need to normalize some paths and also update some documentation surrounding that change.

emesare committed:
Update some tests and examples

emesare committed:
Fix background task tests colliding

We were creating multiple background tasks with the same progress text on multiple threads

emesare committed:
More rust cleanup

- Fixed progress executor freeing itself after one iteration
- Updated the last of the doc imports
- Moved mainthread to main_thread
- Made project creation and opening failable

We could probably AsRef<ProgressExecutor> to get around the allocation and free inside the function bodies, not high priority as those functions are long running anyways.

emesare commented:
I don't know if `rstest` can have fixtures depend on other fixtures, but we need to have the session available at the time of binary view initialization. I _think_ that is what is occurring in CI.

emesare committed:
Move binary view initialization into function body for LLIL test

emesare committed:
Normalize test file names

emesare committed:
More rust cleanup

- Updated README to clarify offline documentation
- Refactored settings api
- Added settings example to show dumping settings value and specific properties
- Use the workspace to depend on binaryninja and binaryninjacore-sys
- Remove binaryninjacore-sys as a workspace member (its not really required)

emesare committed:
Update workflow test to new settings api

emesare committed:
More rust cleanup

- Rename Label to LowLevelILLabel
- Update the functions label map automatically

This fixed a major api blunder where the label map is returned as a reference and originally resulted in UB prone lifetime semantics. It was temporarily "fixed" with explicit label updates in the architecture implementation code. But that was less than ideal and was easy to mess up. Now the label map will be updated automatically as the location of labels is now tracked.

emesare committed:
Misc clippy lints

emesare committed:
More rust cleanup

- Get rid of RawFunctionViewType
- Add better Debug impl for Function

emesare committed:
More rust cleanup

- Fixed the documentation icon using local files (thank you @mkrasnitski)
- Fixed labels being updated and overwriting the label location used to update the label map

emesare commented:
Before this is merged we should add a migration guide somewhere, most likely in the body of the first PR message.

We need to wrap this up this week, there are a few downstream items that depend on this, namely WARP (see: type containers / type archives)

We should put a pin in any further breaking changes and document the modules that are going to be changed in future releases (after next stable).

rbran commented:
One suggestion is to remove the generic `A: Architecture` from the `CallingConvention`: https://github.com/Vector35/binaryninja-api/compare/rust_cleanup_0...rbran:binaryninja-api:cc-core-arch?expand=1

For two main reasons:

First the implementation seems to only use CoreArchitecture by default anyways, so a custom architecture is never specified, adding a complexity with no benefits. Also the `CallingConvention` struct don't make use of the function `BNGetCallingConventionArchitecture`. 

Second to allow the user to define one Custom-CallingConvention and implement it to multiple architectures. The currently implementation only allow one `CallingConvention` (struct) per architecture, because `CallingConventionBase` define the `Arch` as an associated type

rbran commented:
Another good to have feature is to `BinaryViewReader` implementing Read correctly. The [`idb-import`](https://github.com/Vector35/binaryninja-api/blob/da1b04d09832d3a827fd4d97ccdc6818c59b48d8/plugins/idb_import/src/lib.rs#L76) plugin creates a custom implementation of `std::io::Read` for `BinaryView` instead of using `BinaryViewReader` because it don't implement `SeekFrom::End` correctly.

This can be fixed by adding a function that returns the `BinaryView` from the `BinaryViewReader` (or at least the size of it) is added to the core API, like this:

```
pub fn BNGetBinaryReaderView(stream: *mut BNBinaryReader) -> *mut BNBinaryView;
// or at least the size of the view
pub fn BNGetBinaryReaderBinaryViewLen(stream: *mut BNBinaryReader) -> u64;
```

The same applies to `BinaryViewWriter`

emesare commented:
@rbran both of those suggestions seem completely reasonable, I will look into integrating both with this PR. Thank you!

emesare committed:
More rust cleanup

- Added unit tests for MLIL and HLIL
- "Fixed" MLIL, LLIL, and HLIL having issues regarding Instruction vs Expression indexes
- Renamed CallingConvention to CoreCallingConvention and removed architecture generic
- Renamed CallingConventionBase to CallingConvention
- Simplified calling convention code and fixed some bugs with implicit registers
- Added impl Debug to MLIL and HLIL instructions

Still need to at some point add an Expression to MLIL and HLIL. We also might want to look into having the Instruction kind just return the expression kind.

emesare committed:
Misc clippy lint

emesare commented:
Rolling up this PR is gonna be unfortunate from a git blame perspective but that was always going to be the case.

All contiguous "More rust cleanup" commits will most likely be squashed with an incrementing number.

- Rust refactor
- Some commit
- Rust refactor 2
- Another commit
- Rust refactor 3

That or we just squash this PR down to 1 commit. I don't really mind either.

emesare committed:
More rust cleanup

- Allow calling conventions to be registered for multiple architectures
- Swapped a unreachable statement to an unimplemented statement

emesare commented:
@rbran does this look reasonable for reusing a calling convention across architectures?

https://github.com/Vector35/binaryninja-api/pull/6309/commits/b1693fd81fb94c512e8e18da187d273b2cbbef67

emesare commented:
```
test struct_grouper::test_dispatcher_header ... FAILED
test struct_grouper::test_bool_modifier ... FAILED
```

These were failing prior to my changes. However there still seems to be some regression when running through the python tests regarding the PDB import plugin.

rbran commented:
@emesare In my view yes, You implement a CC, then register it for each architecture.

eg: 
```rust
let my_cc = MyCC::new();
let my_cc_x86 = register_calling_convention(arch_x86, "MyCC-x86", my_cc.clone());
let my_cc_arm = register_calling_convention(arch_arm, "MyCC-arm", my_cc);
```

Although it depends on how binja understand the "cc-arch" associations. Maybe it requires a unique CC per arch somewhere inside.

emesare commented:
> Although it depends on how binja understand the "cc-arch" associations. Maybe it requires a unique CC per arch somewhere inside.

The cc provides the register id so as long as the arch keeps the same register id mapping everything should be ok.

All of the same checks are performed that were done prior to this change so while in theory you could receive a bad register mapping, you cannot produce a bad mapping (i.e. rust api will prohibit you from giving the core a bad register id in your CC)

emesare committed:
More rust cleanup

- Fixed the issue with PDB types, this has caused me an insane amount of grief
- Fixed LLIL visit_tree missing LLIL_LOAD expressions
- Added LLIL visitor test
- Made all WARP file pickers use the rfd crate

emesare committed:
Use the dev branch of Binary Ninja in rust CI

emesare committed:
Misc rust fmt

emesare commented:
The last known bug has been fixed with https://github.com/Vector35/binaryninja-api/pull/6309/commits/2c378990dd92d536dc2623d0ae1ab9c51962841c

So the next steps as I see it are:
1. Figure out why the view is failed to be created in CI but not when ran locally
2. Performance testing
3. More documentation?

This PR will be wrapped up by week's end. So if anyone has any go backs now is the last chance.

emesare committed:
More rust cleanup

- Refactored BinaryReader and BinaryWriter
- Added some warnings to high_level_il and medium_level_il modules that they are unstable
- Add BinaryReader and BinaryWriter tests
- Changed BinaryView len to return u64 (that is what the core returns)
- Misc formatting changes
- Remove extern uses in lib.rs

emesare commented:
@rbran I made the changes you suggested in https://github.com/Vector35/binaryninja-api/pull/6309/commits/16e8ad422e7e25d4723869b1264ccae9c5071efb let me know if that covers your use case. I also changed it for the BinaryWriter as well.

emesare committed:
Add impl Debug for BinaryReader and BinaryWriter

emesare committed:
Turn off broken tests

emesare committed:
Add more info to the rust README.md

emesare committed:
More rust cleanup

- Make EdgeStyle type not wrap raw
- Regression tests for WARP will run on all bins in the out dir now

emesare committed:
Update collaboration API

Makes collaboration more in line with the refactor. Still a lot of work to do.

Namely still need:
- Proper errors
- _with_opts functions
- More ergonomic api
- Better connection procedure
- Updated documentation
- A LOT of unit tests
- An example
- Typed id's for everything (i dont want BnString as the id!!!)
- NEED to refactor the progress callbacks into the new progress api, but we should pull in some of the stuff the collab progress has
- Elimination of apis that are dumb helpers

emesare commented:
Because I have tomorrow and Friday to finish this up I went ahead and pulled in https://github.com/Vector35/binaryninja-api/pull/5773, I think I should have enough time to refactor it and add some tests and examples.

rbran commented:
@emesare Thanks that solves my use case.

BTW I hope in the future we can add a function to the coreAPI to recover the `BinaryView` from a `BinaryReader`/`BinaryWriter`, this allow rust to implement `#[repr(transparent)]` for `BinaryReader/Writer` to `BNBinaryReader/Writer`. But that's a task for other PR.

emesare committed:
Separate out the rust testing and use pull_request_target

pull_request_target allows PR's to access the headless license, for this to be safe we need to prevent people from running the job.

To prevent the job from being ran we add an environment requirement on testing that a reviewer must review the code and then manually approve it to run.

emesare committed:
More rust cleanup

- Use GroupId instead of u64
- Use ProgressCallback in place of ProgressExecutor
- Misc cleanup of FileMetadata
- Add `save_to_path` and `save_to_accessor` to save modified binaries
- Added binary_view unit tests
- Added collaboration unit tests
- Fixed a few issues with the collaboration apis
- Renamed Command registration functions so that there is no import ambiguity
- Split out RemoteUndoEntry
- Collaboration apis now have a explicit `_with_progress` set of apis
- Misc clippy lint fixes

emesare committed:
Fix some typos

emesare committed:
More rust cleanup

- Add extra info to README.md
- Refactor components api
- Add components unit test

emesare commented:
The rolled up commit will be on https://github.com/Vector35/binaryninja-api/tree/rust_refactor

emesare commented:
Community rust plugins to update:
- https://github.com/starfleetcadet75/binaryninja-superh
- https://github.com/unknowntrojan/binja_coolsigmaker

emesare commented:
Merged with https://github.com/Vector35/binaryninja-api/commit/6a63c17853009ce7a0071458e39c140a09440230

blacktop commented:
@emesare are you going to add a load with progress func for the headless session?

emesare commented:
> @emesare are you going to add a load with progress func for the headless session?

That sounds like a good idea, I'll add that today! Thank you.

emesare commented:
@blacktop added with https://github.com/Vector35/binaryninja-api/commit/98276031969e67229d6cc6344f1b9debdc211e8c

I should note that it appears the core only calls that progress callback for loading BNDBs, IIRC there was a comment somewhere about that in the rust code. I could not find that comment so I added it to the respective functions with https://github.com/Vector35/binaryninja-api/commit/0263957dc920a10785af6874ab3040a529547d70

blacktop commented:
Thank you! 

I know this is most likely an inappropriate place to ask, but I'm trying to understand how to run a plugin via the Rust API (BinExport) specifically.  Are there any examples anywhere on how to do this? I can also ask on Slack if that's a better place?

emesare commented:
No problem, by run you mean run a command that BinExport registered?

Like one of these:
<img width="704" alt="image" src="https://github.com/user-attachments/assets/171a26e7-90fe-4464-9487-ea47822927b1" />

If so than we would need to add the rust equivalent of `PluginCommand` https://github.com/Vector35/binaryninja-api/blob/8767400b712f12f459556844e27d9b9f5ff037bd/binaryninjaapi.h#L14146-L14168

In rust we can register commands but we cannot call them yet.

I unfortunately have a few things in front of adding that, specifically I need to implement the BinaryDataNotification API. If you would like to make an issue for fleshing out plugin commands to support calling them from rust, I can assign it to myself.

emesare commented:
Also this is assuming the plugin command that you want to call does not then invoke some user interaction, that would require us to also add the interaction callback API.

There is a TODO for that in the flow graph example https://github.com/Vector35/binaryninja-api/blob/6a63c17853009ce7a0071458e39c140a09440230/rust/examples/flowgraph.rs#L52-L53

blacktop commented:
@emesare I guess this is how it currently works in python? https://infosec.exchange/@binaryninja/113827475796582876

blacktop commented:
> No problem, by run you mean run a command that BinExport registered?
> 
> Like one of these: <img alt="image" width="704" src="https://private-user-images.githubusercontent.com/35282038/406690566-171a26e7-90fe-4464-9487-ea47822927b1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc4NDMxNjEsIm5iZiI6MTczNzg0Mjg2MSwicGF0aCI6Ii8zNTI4MjAzOC80MDY2OTA1NjYtMTcxYTI2ZTctOTBmZS00NDY0LTk0ODctZWE0NzgyMjkyN2IxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTI1VDIyMDc0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc5ODE1NGQyOWRhMTE4Y2I1OWI2YjM5MGM2MmYyNDBhYmJiM2ZkMWQwMWVjMmNhYmEyYzk5ZmVkNjIyZDg3ZDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.O0uI2ZG-acSo-sccfKJzVrpPJi8UlzRJEOI8mk72RK0">
> 
> If so than we would need to add the rust equivalent of `PluginCommand`
> 
> https://github.com/Vector35/binaryninja-api/blob/8767400b712f12f459556844e27d9b9f5ff037bd/binaryninjaapi.h#L14146-L14168
> 
> In rust we can register commands but we cannot call them yet.
> 
> I unfortunately have a few things in front of adding that, specifically I need to implement the BinaryDataNotification API. If you would like to make an issue for fleshing out plugin commands to support calling them from rust, I can assign it to myself.

Ah I see, thank you for this explanation. I will wait for PluginCommand running to be added to the Rust API. Thanks again!

emesare commented:
Yea that would be through the `PluginCommand`, if this is something you need now than you can call `BNGetAllPluginCommands` yourself until I or someone else can get to it.

Here are the issues:
- https://github.com/Vector35/binaryninja-api/issues/6362
- https://github.com/Vector35/binaryninja-api/issues/6361 (not required to solve your use case)
